flocker/node/agents/test/test_blockdevice.py

```diff
@@ -1270,37 +1414,116 @@ def test_simple_transitions(self, two_dataset_states):
+                unicode(dataset_id))
+            if getattr(initial_dataset,
+                       'mount_point',
+                       _NoSuchThing) is not _NoSuchThing:
+                initial_dataset = initial_dataset.set(mount_point=mount_point)
+            if getattr(next_dataset,
+                       'mount_point',
+                       _NoSuchThing) is not _NoSuchThing:
+                next_dataset = next_dataset.set(mount_point=mount_point)
+
+            # Converge to the initial state.
+            try:
+                test_objects.run_to_convergence([initial_dataset])
+            except DidNotConverge as e:
+                self.fail(
+                    "Did not converge to initial state after %d iterations." %
+                    e.iteration_count)
+
+            # Converge from the initial state to the next state.
+            try:
+                test_objects.run_to_convergence([next_dataset])
+            except DidNotConverge as e:
+                self.fail("Did not converge to next state after %d "
+                          "iterations." % e.iteration_count)
+        finally:
+            test_objects.cleanup_example()
+
+    @given(
+        two_dataset_states=TWO_DESIRED_DATASET_STRATEGY
+    )
+    def test_docker_protection(self, two_dataset_states):
+        """
+        Given an initial empty state, ``BlockDeviceCalculator`` will converge
+        to any ``DesiredDataset``, followed by any other state of the same
+        dataset. During that time, it will never be the case that a
+        successfully write to a mount path that has been sent to the control
+        agent will not be persisted to the backing volume.
+
+        This would represent an agent like docker attempting to use a path
+        before or after flocker was ready for it to use the path.
+        """
         initial_dataset, next_dataset = two_dataset_states
 
         dataset_id = initial_dataset.dataset_id
 
-        # Set the mountpoint to a real mountpoint in desired dataset states
-        # that have a mount point attribute.
-        mount_point = self.deployer._mountpath_for_dataset_id(
-            unicode(dataset_id))
-        if getattr(initial_dataset,
-                   'mount_point',
-                   _NoSuchThing) is not _NoSuchThing:
-            initial_dataset = initial_dataset.set(mount_point=mount_point)
-        if getattr(next_dataset,
-                   'mount_point',
-                   _NoSuchThing) is not _NoSuchThing:
-            next_dataset = next_dataset.set(mount_point=mount_point)
-
-        # Converge to the initial state.
-        try:
-            self.run_to_convergence([initial_dataset])
-        except DidNotConverge as e:
-            self.fail(
-                "Did not converge to initial state after %d iterations." %
-                e.iteration_count)
+        evaluate_function = None
+
+        def nullable_callback():
```

Comment:
:cry:

I suggest a class with `__call__`. As it is I've no confidence `evaluate_function` will ever be called, scoping rules are too tricksy...